pf/internal/convert/encoding.go

```diff
@@ -253,7 +253,17 @@ func (e *encoding) deriveEncoder(typeSpec *pschema.TypeSpec, t tftypes.Type) (En
 		case tftypes.Tuple:
 			return e.deriveTupleEncoder(typeSpec.Ref, t)
 		default:
-			return nil, fmt.Errorf("expected Object or Tuple type but got %s", t.String())
+			// Workaround dangling local references.
+			dangling := false
+			if strings.HasPrefix(typeSpec.Ref, "#/types/") {
+				token := strings.TrimPrefix(typeSpec.Ref, "#/types/")
+				if e.spec.Type(tokens.Type(token)) == nil {
+					dangling = true
+				}
+			}
+			if !dangling {
+				return nil, fmt.Errorf("expected Object or Tuple type but got %s", t.String())
+			}
```

Comment:
Schema types with `Ref` fields that refer to non-existent types are called ["token types"](https://github.com/pulumi/pulumi/blob/6dd2b026916c2e999502100ae16f47faf3737690/pkg/codegen/schema/schema.go#L282-L289). For the purposes of deriving an encoding, they are equivalent to their underlying type. 

Instead of handling this in the `default` case (and causing `tftypes.Object` and `tftypes.Tuple` to fail with an unhelpful error in this case), I think we should check if the `Ref` is dangling before we enter the `switch`:

```go
if typeSpec.Ref != "" && e.validRef(typeSpec.Ref) {
    switch t := t.(type) {
        ...
    }
}
```

Reply:
Hm that's fair. We're doing co-recursion over tftypes.Type and pschema.TypeSpec which is always a little odd to write; we could perhaps benefit from pattern-matching on tftypes.Type first. I took a pass simplifying this so we start the object/tuple cases by looking at tftypes.Type. IFF that needs an object or tuple, we require a local resolvable "ref". This seems reasonable for now. Leaving non-local refs for later.